main.vala

```diff
@@ -639,291 +522,1662 @@ class Vls.Server {
                 sb.insert ((ssize_t) pos, changeEvent.text);
             }
         }
-        source.file.content = sb.str;
+        source.content = sb.str;
 
-        // if we're at this point, the file is present in the context
-        // any change we make invalidates the context
-        ctx.invalidate ();
+        source.compilation.invalidate ();
+        request_context_update (client);
+    }
 
-        // we have to update everything
-        ctx.check ();
+    void request_context_update (Jsonrpc.Client client) {
+        update_context_client = client;
+        update_context_requests += 1;
+        int64 delay_us = int64.min (update_context_delay_inc_us * update_context_requests, update_context_delay_max_us);
```

Comment:
>update_context_delay_inc_us * update_context_requests
this increases the delay as the number of requests increase, is that the intention?

also, it looks like the worst possible case for a context update is actually 1 second, right?

Reply:
> this increases the delay as the number of requests increase, is that the intention?

Yes.

> also, it looks like the worst possible case for a context update is actually 1 second, right?

The worst-possible delay that we've set, yes. The effective delay could be much longer depending on how long compilation takes.